File: crates/peakd/js/bench.js
/**
 * In-app frame benchmark, injected by the browser only when `--benchmark` is
 * passed. It never runs in a normal launch.
 *
 * Why it lives in the page rather than in an external harness: every attempt to
 * drive this from outside (synthetic CGEvents, Accessibility window moves, an
 * AppleScript bridge) turned into a fight with macOS window occlusion, because
 * a background or occluded WKWebView has `requestAnimationFrame` throttled to a
 * near stop. Measuring from inside the page, in the process that actually
 * paints, removes that entire class of flakiness.
 *
 * What it measures, and what it deliberately does not:
 *   - It counts `requestAnimationFrame` callbacks, which is the rate the
 *     compositor is actually presenting frames at. The app's own animation
 *     shares this cadence, so the number is the browser's, not a private loop's.
 *   - It is *passive*: it never touches the app's DOM, styles or timers. The
 *     only global it creates is `__peakdBench`, so it cannot be mistaken for
 *     app state.
 *   - It records window moves and resizes as they happen, so a slide of the
 *     window during the run is visible in the report rather than averaged away.
 *
 * Collapse in the report is the thing to look at, not the mean: a mean hides a
 * 200 ms stall behind a thousand good frames, and a stall is exactly what a
 * user perceives as "it stutters when I drag it fast".
 */
(function () {
  "use strict";

  if (window.__peakdBench) return; // never double-instrument

  var bench = {
    running: true,
    startedAt: 0,
    frames: 0,
    worstFrameMs: 0,
    worstFrameAtMs: 0,
    /** Frames slower than 32 ms (a 60 Hz budget) and 50 ms (visibly janky). */
    over32: 0,
    over50: 0,
    /** Every inter-frame gap above 32 ms, with when it happened. */
    stalls: [],
    /** Window movement, as seen by the page (viewport resizes). */
    resizes: [],
    longTasks: [],
    started: false,
    durationMs: 0,
  };

  window.__peakdBench = bench;

  // Long tasks are the other half of the story: a frame can be late because a
  // task held the main thread, which shows up here rather than in the frame
  // deltas.
  try {
    if (window.PerformanceObserver && PerformanceObserver.supportedEntryTypes &&
        PerformanceObserver.supportedEntryTypes.indexOf("longtask") !== -1) {
      new PerformanceObserver(function (list) {
        list.getEntries().forEach(function (e) {
          bench.longTasks.push({
            atMs: Math.round(e.startTime - bench.startedAt),
            ms: Math.round(e.duration),
          });
        });
      }).observe({ entryTypes: ["longtask"] });
    }
  } catch (e) {}

  // A resize during the run is a strong hint that whoever is measuring is
  // dragging the window; recording it lets the report separate "slow while
  // moving" from "slow in general".
  window.addEventListener("resize", function () {
    if (!bench.started || !bench.running) return;
    var now = performance.now();
    bench.resizes.push({
      atMs: Math.round(now - bench.startedAt),
      w: window.innerWidth,
      h: window.innerHeight,
    });
  });

  var last = 0;

  function tick(now) {
    if (!bench.running) return;
    if (!bench.started) {
      // The first callback establishes the clock; it is not a real frame.
      bench.started = true;
      bench.startedAt = now;
      last = now;
      window.requestAnimationFrame(tick);

      return;
    }

    var dt = now - last;
    last = now;
    bench.frames++;
    bench.durationMs = now - bench.startedAt;

    if (dt > bench.worstFrameMs) {
      bench.worstFrameMs = dt;
      bench.worstFrameAtMs = Math.round(now - bench.startedAt);
    }
    // Ignore the first couple of frames after start: the clock is settling.
    if (bench.frames > 3) {
      if (dt > 32) bench.over32++;
      if (dt > 50) {
        bench.over50++;
        if (bench.stalls.length < 50) {
          bench.stalls.push({
            atMs: Math.round(now - bench.startedAt),
            ms: Math.round(dt),
          });
        }
      }
    }

    window.requestAnimationFrame(tick);

  }

  window.requestAnimationFrame(tick);


  /** Snapshot for the host to collect. */
  window.__peakdBenchReport = function () {
    var b = window.__peakdBench;
    var seconds = b.durationMs / 1000;
    return {
      durationMs: Math.round(b.durationMs),
      frames: b.frames,
      fps: seconds > 0 ? +(b.frames / seconds).toFixed(1) : 0,
      worstFrameMs: +b.worstFrameMs.toFixed(1),
      worstFrameAtMs: b.worstFrameAtMs,
      framesOver32ms: b.over32,
      framesOver50ms: b.over50,
      stalls: b.stalls,
      resizes: b.resizes,
      longTasks: b.longTasks,
      // Context so a result is comparable across machines.
      dpr: window.devicePixelRatio,
      viewport: { w: window.innerWidth, h: window.innerHeight },
      canvases: document.querySelectorAll("canvas").length,
      tiles: document.querySelectorAll("#tile-grid .tile").length,
      nodes: document.querySelectorAll("*").length,
    };
  };
})();

File: crates/peakd/src/bench.rs
//! In-app frame benchmark for the Qt shell.
//!
//! The page-side instrumentation is the same as the GTK shell's (`js/bench.js`,
//! temporarily copied from `crates/peakd/js/`) and the report format is
//! identical, so kiosk numbers are comparable across the two shells.
//!
//! Scheduling lives on the shell's 100 ms pump instead of a tao event loop:
//! `tick` advances a small state machine (settle → inject → measure →
//! collect), and window moves arrive through the shim's `window`/`move` view
//! event, which is the drag signal the report labels "moving".

use std::io::Write;
use std::time::{Duration, Instant};

use serde_json::Value;

use crate::config::Benchmark;

/// The page-side instrumentation, compiled in so there is no path to get wrong
/// at runtime.
pub const BENCH_JS: &str = include_str!("../js/bench.js");

/// Synthetic drag of a floating plugin window, used by `--benchmark-drag-tile`.
pub const DRAG_TILE_JS: &str = include_str!("../js/bench-drag-tile.js");

/// What the pump should do this tick.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Step {
    /// Inject the instrumentation (and the tile drag when configured).
    Inject,
    /// Read the report back.
    Collect,
}

#[derive(Clone, Copy)]
enum Phase {
    Waiting { until: Instant },
    Measuring { started: Instant, collect_at: Instant },
    Collecting,
    Done,
}

pub struct BenchState {
    config: Benchmark,
    drag_tile: bool,
    phase: Phase,
    moves: u64,
    first_move_ms: Option<u64>,
}

impl BenchState {
    pub fn new(config: Benchmark, drag_tile: bool) -> Self {
        let until = Instant::now() + Duration::from_secs(config.settle_seconds);
        Self {
            config,
            drag_tile,
            phase: Phase::Waiting { until },
            moves: 0,
            first_move_ms: None,
        }
    }

    #[must_use]
    pub fn drag_tile(&self) -> bool {
        self.drag_tile
    }

    /// Advance the schedule; `None` means "nothing to do this tick".
    pub fn tick(&mut self) -> Option<Step> {
        match self.phase {
            Phase::Waiting { until } if Instant::now() >= until => {
                let started = Instant::now();
                self.phase = Phase::Measuring {
                    started,
                    // Margin so the page has closed its measurement window
                    // before the report is read; reading early truncates it.
                    collect_at: started + Duration::from_millis(self.config.seconds * 1000 + 1200),
                };
                Some(Step::Inject)
            }
            Phase::Measuring { collect_at, .. } if Instant::now() >= collect_at => {
                self.phase = Phase::Collecting;
                Some(Step::Collect)
            }
            _ => None,
        }
    }

    /// Record a window move. The first one, relative to the measurement start,
    /// is what makes the report's "moving" case honest.
    pub fn observe_window_move(&mut self) {
        if let Phase::Measuring { started, .. } = self.phase {
            self.moves += 1;
            if self.first_move_ms.is_none() {
                self.first_move_ms = Some(started.elapsed().as_millis() as u64);
            }
        }
    }

    /// Parse, annotate, print and optionally write the report.
    pub fn finish(&mut self, raw: &str) {
        self.phase = Phase::Done;
        finish(raw, self.moves, self.first_move_ms, self.config.output.as_deref());
    }
}

pub fn announce(config: &Benchmark) {
    eprintln!(
        "benchmark: settling {}s, then measuring {}s",
        config.settle_seconds, config.seconds
    );
    eprintln!("benchmark: drag the window while it measures to get the moving case");
}

/// Turns the raw counters into the one thing a reader wants: which mechanism
/// stalled, if any.
///
/// The comparison is between two independent clocks. `requestAnimationFrame`
/// and `setInterval` are throttled by *different* rules, so:
///
/// * rAF stalls, timers keep ticking  -> the page is being *frame* throttled.
/// * rAF and timers both stall        -> the main thread is starved.
/// * neither stalls, yet it feels slow -> the cost is outside the page
///   entirely: the compositor not taking frames.
fn diagnose(obj: &serde_json::Map<String, Value>) -> Value {
    let fps = obj.get("fps").and_then(Value::as_f64).unwrap_or(0.0);
    let seconds = obj
        .get("durationMs")
        .and_then(Value::as_f64)
        .unwrap_or(0.0)
        / 1000.0;
    let timer_ticks = obj.get("timerTicks").and_then(Value::as_u64).unwrap_or(0);
    let timer_worst = obj
        .get("timerWorstMs")
        .and_then(Value::as_f64)
        .unwrap_or(0.0);
    let worst = obj.get("worstFrameMs").and_then(Value::as_f64).unwrap_or(0.0);
    let stalls = obj
        .get("stalls")
        .and_then(Value::as_array)
        .map(|a| a.len())
        .unwrap_or(0);
    let hidden = obj
        .get("hiddenAtStart")
        .and_then(Value::as_bool)
        .unwrap_or(false);

    // The canary fires every 100 ms, so this is the share of expected ticks.
    let expected_ticks = (seconds * 10.0).max(1.0);
    let timer_share = (timer_ticks as f64 / expected_ticks).min(1.0);

    let starved_detail = format!(
        "the timer canary stalled too (worst gap {timer_worst:.0} ms, {:.0}% of expected \
         ticks), so the main thread itself was blocked rather than frames being \
         withheld. Look at longTasks and stalls for the cause",
        timer_share * 100.0
    );

    let (verdict, detail) = if hidden {
        (
            "page-hidden",
            "the page reported document.hidden at the start: the engine throttles \
             requestAnimationFrame in a hidden or occluded page, so nothing here \
             describes a visible window",
        )
    } else if timer_share < 0.85 || timer_worst > 250.0 {
        ("main-thread-starved", starved_detail.as_str())
    } else if fps > 0.0 && worst <= 40.0 && stalls == 0 {
        (
            "no-page-side-stall",
            "frames were produced on schedule and no stall exceeded 50 ms, so the \
             page and its main thread were healthy for this window. If dragging \
             still felt slow, the cost is outside the page — the compositor \
             recompositing a moving window — which JavaScript cannot observe",
        )
    } else if stalls > 0 {
        (
            "frames-withheld",
            "requestAnimationFrame was withheld while the timer canary kept running: \
             frame throttling, not main-thread starvation",
        )
    } else {
        (
            "inconclusive",
            "no single mechanism dominates; read worstFrameMs and stalls directly",
        )
    };

    // Bind first: a leading unary `+` is a JS habit, not Rust, and is not valid
    // where a JSON macro expects an expression.
    let timer_share_pct = (timer_share * 100.0).round() / 100.0;
    serde_json::json!({
        "verdict": verdict,
        "detail": detail,
        "timerShare": timer_share_pct,
    })
}

/// Normalises, annotates, prints and optionally writes the report.
fn finish(raw: &str, moves: u64, first_move_ms: Option<u64>, output: Option<&str>) {
    let json = raw.trim().trim_matches('"').replace("\\\"", "\"");
    let mut value: Value = serde_json::from_str(&json).unwrap_or_else(|_| {
        serde_json::json!({ "error": "could not parse the report", "raw": json })
    });

    if let Some(obj) = value.as_object_mut() {
        obj.insert("windowMoves".into(), serde_json::json!(moves));
        if let Some(at) = first_move_ms {
            obj.insert("firstMoveAtMs".into(), serde_json::json!(at));
        }
        // State the verdict in the report so it cannot be quietly misread: a
        // run with no window movement is the idle case and nothing else.
        let fps = obj.get("fps").and_then(Value::as_f64).unwrap_or(0.0);
        let worst = obj.get("worstFrameMs").and_then(Value::as_f64).unwrap_or(0.0);
        obj.insert(
            "verdict".into(),
            serde_json::json!({
                "fps": fps,
                "worstFrameMs": worst,
                "case": if moves > 0 { "moving" } else { "idle" },
                "note": if moves > 0 {
                    "the window moved during the run, so this includes the moving case"
                } else {
                    "the window never moved: this is the idle case only"
                },
            }),
        );
    }

    if let Some(obj) = value.as_object_mut() {
        obj.insert("diagnosis".into(), diagnose(obj));
    }

    let pretty = serde_json::to_string_pretty(&value).unwrap_or(json);
    println!("{pretty}");
    // Flush before exiting: with stdout piped, an unflushed buffer is lost.
    let _ = std::io::stdout().flush();

    if let Some(path) = output {
        match std::fs::write(path, format!("{pretty}\n")) {
            Ok(()) => eprintln!("benchmark: report written to {path}"),
            Err(err) => eprintln!("benchmark: could not write {path}: {err}"),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn diagnose_json(raw: &str) -> Value {
        let value: Value = serde_json::from_str(raw).unwrap();
        diagnose(value.as_object().unwrap())
    }

    #[test]
    fn healthy_page_is_not_stalled() {
        let verdict = diagnose_json(
            r#"{"fps":30,"durationMs":10000,"timerTicks":100,"timerWorstMs":120,"worstFrameMs":20,"stalls":[]}"#,
        );
        assert_eq!(verdict["verdict"], "no-page-side-stall");
    }

    #[test]
    fn hidden_page_is_called_out() {
        let verdict = diagnose_json(r#"{"hiddenAtStart":true,"fps":0}"#);
        assert_eq!(verdict["verdict"], "page-hidden");
    }

    #[test]
    fn stalled_timer_means_starved_main_thread() {
        let verdict = diagnose_json(
            r#"{"fps":5,"durationMs":10000,"timerTicks":10,"timerWorstMs":900,"worstFrameMs":80,"stalls":[]}"#,
        );
        assert_eq!(verdict["verdict"], "main-thread-starved");
    }

    #[test]
    fn window_moves_mark_the_moving_case() {
        let mut state = BenchState::new(
            Benchmark {
                seconds: 1,
                settle_seconds: 0,
                output: None,
            },
            false,
        );
        // Settle is zero, so the first tick starts measuring.
        assert_eq!(state.tick(), Some(Step::Inject));
        state.observe_window_move();
        state.observe_window_move();
        assert_eq!(state.moves, 2);
        assert!(state.first_move_ms.is_some());
        // Still inside the measurement window: nothing to collect yet.
        assert_eq!(state.tick(), None);
    }
}
